_blog/2023-09-13-building-web-scraper-in-rust.mdx

```diff
@@ -0,0 +1,381 @@
+            let url = format!("https://www.amazon.com/s?k=raspberry+pi&page={pagenum}");
+
+            let res = match ctx.get(url).send().await {
+                Ok(res) => res,
+                Err(e) => {
+                    error!("Something went wrong while fetching from url: {e}");
+                    StdSleep(StdDuration::from_secs(15));
+                    continue;
+                }
+            };
+
+            if res.status() == StatusCode::SERVICE_UNAVAILABLE {
+                error!("Amazon returned a 503 at page {pagenum}");
+                retry_attempts += 1;
+                if retry_attempts >= 10 {
+                    error!("It looks like Amazon is blocking us! We will rest for an hour.");
+                    StdSleep(StdDuration::from_secs(3600));
+                    continue;
+                } else {
+                    StdSleep(StdDuration::from_secs(15));
+                    continue;
+                }
+            }
+
+            let body = match res.text().await {
+                Ok(res) => res,
+                Err(e) => {
+                    error!("Something went wrong while turning data to text: {e}");
+                    StdSleep(StdDuration::from_secs(15));
+                    continue;
+                }
+            };
+
+            debug!("Page {pagenum} was scraped");
+            let html = Html::parse_fragment(&body);
+            let selector =
+                Selector::parse("div[data-component-type= ' s-search-result ' ]").unwrap();
+
+            if html.select(&selector).count() == 0 {
+                break;
+            };
+
+            for entry in html.select(&selector) {
+                let price_selector = Selector::parse("span.a-price > span.a-offscreen").unwrap();
+                let productname_selector = Selector::parse("h2 > a").unwrap();
+
+                let price_text = entry
+                    .select(&price_selector)
+                    .map(|x| x.text().next().unwrap().to_string())
+                    .collect::<Vec<String>>();
+
+                vec.push(Product {
+                    name: entry
+                        .select(&productname_selector)
+                        .next()
+                        .expect("Couldn't find the product name!")
+                        .text()
+                        .next()
+                        .unwrap()
+                        .to_string(),
+                    price: price_text[0].clone(),
+                    old_price: Some(price_text[1].clone()),
+                    link: entry
+                        .select(&productname_selector)
+                        .map(|link| {
+                            format!("https://amazon.co.uk{}", link.value().attr("href").unwrap())
+                        })
+                        .collect::<String>(),
+                });
+            }
+            pagenum += 1;
+                retry_attempts = 0;
+            StdSleep(StdDuration::from_secs(15));
```

Comment:
btw, I see no reason for why this could not be a tokio sleep. But it don't matter much, it's essentially a sequential program.

Reply:
amended in next commit

I was having some issues with `scraper`-related `!Send` errors, so this was overlooked